### ISSUE TITLE
Remove vote dust threshold

### ISSUE THREAD

**Issue by mvandeberg** [steemit/steem#1764]
We are removing the vote dust threshold in HF 20.

We have two options in how to implement this. 

1. Do the vote as is
2. Floor the vote to 0 rshares

I am inclined towards 1. The vote dust threshold was instated during quadratic rewards. Votes under the threshold could add up to decent reward and should not be overlooked.

**Comment by TimCliff** [comment]
I would much prefer option 2 (Floor the vote to 0 rshares), as it will discourage users from making lots of small dust votes which increases voting spam. If a user wants to make a 'dust vote', it would still be allowed and counted (so the user gets to express their like) but it will not contribute to the rewards until the user acquires more SP.

One other thing that would be good to consider along with this issue would be to apply the "floor of dust votes" shift amount to **all** votes. This would provide a (very small) disincentive for users to split their votes across multiple accounts, and would slightly favor users who consolidated their voting into a single account - which would also help to reduce vote spam.

A third thing to consider (which would be best to discuss in a separate issue if it is accepted) is to remove the limit on being able to power down an account, so that users who have lots of small dust accounts can power them down and consolidate.

**Comment by iamsmooth** [comment]
Another suggestion that is related is to impose a fixed deduction/cost per vote, even for votes above the threshold. This will slightly disadvantage the smallest SP accounts with a degree of sub-linearity, but also discourage spam/dust voting and encourage consolidating stake into a smaller number of accounts/votes, at least to the extent it is reward-motivated (also may serve to increase incentives for the smallest SP accounts to power up). I still basically agree with the original comment here that users should be able to make zero rshares votes for social and UX purposes (subject to other bandwidth/spam limits), but maybe these should be implemented in some non-consensus way?

**Comment by mvandeberg** [comment]
> This will slightly disadvantage the smallest SP accounts with a degree of sub-linearity

It is a slight degree of super-linearity. Not mathematically speaking, of course. If you were to graph the SP to rshare ratio, then it ends up being slightly super-linear with a linear limit.

**Comment by mvandeberg** [comment]
I had made up some similar graphs when working on previous curve changes. Can't remember if I shared them or not. The curve itself starts at 0 and convexly increases to a limit of 1. An increasing function tends towards super-linearity and a decreasing function tends towards sub-linearity. For comparison, the graph for r^2 rewards would be r. (`r^2/r`). The form we have hear is (`r/(r-N)`). Which is the base form that the previously proposed curve was derived from, so we may be able to conclude that the previously proposed curve and this curve belong to some family of curves with this ratio form.

If we make this change we should probably perform the curation derivation again. But I have a hunch that the constant will disappear and we will have the same curation curve that we do now.

**Comment by iamsmooth** [comment]
> It is a slight degree of super-linearity. Not mathematically speaking, of course

You are of course correct looking at it from the perspective of the curve itself. I was referring to how the small account user would see it: 'sub-linear' in the sense that their vote influence is 'below' what it would be in a linear system. That's confusing terminology though, as we could say the same for a true super-linear curve.

**Comment by theoreticalbts** [comment]
> If we make this change we should probably perform the curation derivation again.

You're getting confused.  I suggest re-reading the SMT whitepaper section "Content rewards", it has the best explanation of the rewards system I've come up with.  In Fig 8, the horizontal axis is R-shares.  So if we decide some upvotes don't generate R-shares, ipso facto those upvotes don't generate curation rewards either.

A *per-post dust threshold* would be equivalent to shifting the reward and weight curves, adding a section at the front where it's zero.  But we've never talked about a per-post dust threshold.

In this ticket, we're talking about a *per-author dust threshold*, which is basically a response function which modulates the amount by which a given voter's upvote moves from left-to-right along the reward and weight curves.  Currently this response function is implicitly the identity function `y = x`, or using more verbose descriptive variable names, such as `effective_rshares = rshares`.

> apply the "floor of dust votes" shift amount to all votes. This would provide a (very small) disincentive for users who split their votes across multiple accounts, and would slightly favor users who consolidated their voting into a single account - which would also help to reduce vote spam

I like this idea.  You can think of this as having a "paid" membership tier for having your upvote influence rewards, the "fee" for the membership tier is keeping enough VESTS to overcome the threshold, which don't count in your votes.  Unlike a more traditional idea of paid membership tiers, with this system you can "cancel" your membership and get your "fee" back with a full power down.  

Here are the options as I see them:

```
// (1) current behavior
effective_rshares = rshares;

// (2) original conception of this change
effective_rshares = (rshares > dust_threshold) ? rshares : 0;

// (3) iamsmooth's suggestion
effective_rshares = max(0, rshares-dust_threshold);

// (4) combine them both
effective_rshares = (rshares > dust_threshold) ? rshares-dust_threshold/2 : 0;
```

> keeping enough VESTS to overcome the threshold, which don't count in your votes

Actually with option (4), effectively it counts 50%.

Personally, I like option (4) since it achieves both objectives:

- (a) Don't process votes if they wouldn't affect rewards very much anyway (2, 4)
- (b) Slight penalty for splitting funds up among multiple accounts (3, 4)

We can implement any of the above options by performing the assignment at some point early in processing, then replacing all subsequent references to `rshares` with `effective_rshares`.  We should probably have one commit that implements option (1), so if you read the commit it's obvious the behavior is equivalent.  Then have another commit that switches from option (1) to whichever of the above we pick, it's a one-liner (well, some values in the tests may be wrong and need to be changed).

`dust_threshold` should probably be a witness parameter, because it may change with economics.  Fortunately now there's code which allows us to easily add new witness parameters.

**Comment by mvandeberg** [comment]
How did you come up with a halving the threshold? Why not a third or a quarter? It seems to be an extra variable added to create a more generic solution.

The following behavior should be noted:

Let's call the divisor, the `effective_dust_weight` as it is how much weight the dust threshold has in the effective rshares for any vote with `abs_rshares > 0`.

`lim effective_dust_weight -> 0 = (2)`
`lim effective_dust_weight -> 1 = (3)`

So we actually have a gradient of behaviors between (2) and (3) that can be used.

The current branch implements (3).

**Comment by theoreticalbts** [comment]
> How did you come up with a halving the threshold? Why not a third or a quarter?

You're right, there are actually two thresholds here:

- rshares below `T1` don't count.
- If you have rshares above `T1`, they *still* don't count until you have at least `T2` rshares above `T1`, then all the rshares above `T1` start counting at once.

which may be expressed in code as:

```
effective_rshares = (rshares > T1+T2) ? rshares-T1 : 0;
```

We haven't actually explicitly outlined the requirements, but implicitly I sort of assumed they are:

- `dust_threshold` is a value that is near the largest economically meaningless amount of rshares.
- `T1` should not be an economically meaningful amount of rshares; subject to that constraint, larger `T1` is better, so we should set `T1 = dust_threshold`.
- `T2` should not be an economically meaningful amount of rshares; subject to that constraint, larger `T2` is better, so we should set `T2 = dust_threshold`.

Substituting we get

```
effective_rshares = (rshares > 2*dust_threshold) ? rshares-dust_threshold : 0;
```

It's off by a factor of 2, but if we tweak our definitions a little bit, for example if we say

- `dust_threshold` is a value that is nearly twice the largest economically meaningless amount of rshares

and re-do the above analysis, we find that it matches what I wrote above.

**Comment by theoreticalbts** [comment]
You can think of why `T1` should be equal to `T2` in terms of switchover points:

- If `T1` is too small, we can increase it and get more value by disallowing upvotes from a larger set of spammy accounts that have tiny balances and wouldn't affect the outcome much anyway.
- If `T1` is too large, we can decrease it to stop denying influence over voting rewards to legitimate users who have spent economically meaningful amounts of money increasing their SP.
- At some point `T1` will "switchover" from being too small to being too large, that's the "Goldilocks value of `T1`" -- not too small, not too large, just right!

- If `T2` is too small, we can increase it and get more value by not performing calculations involving very small amounts of `rshares`.
- If `T2` is too large, we can decrease it to stop denying influence over voting rewards to legitimate users who have spent economically meaningful amounts of money increasing their SP.
- At some point `T2` will "switchover" from being too small to being too large, that's the "Goldilocks value of `T2`" -- not too small, not too large, just right!

The switchover point / Goldilocks value should be the same for `T1` and `T2` because they're both based around similar considerations of "what is the cutoff for between economically meaningless and economically meaningful?"

**Comment by theoreticalbts** [comment]
How should we define the witness parameter called `vote_dust_threshold`?

- (VDT1) `vote_dust_threshold` should be defined as `vote_dust_threshold = T1 = T2`
- (VDT2) `vote_dust_threshold` should be defined as `vote_dust_threshold = T1+T2 = 2*T1 = 2*T2`

It doesn't matter whether we choose VDT1 or VDT2.  If we implement VDT2 the witnesses will just set the parameter to twice what they would have set it to if we had implemented VDT1.

**Comment by mvandeberg** [comment]
This is expanding far beyond the initial issue especially if we make the parameter witness votable. Not necessarily a bad thing, just more than what was planned for this change. And requiring significantly more testing.

I am hesitant to make the threshold witness votable because that would mean effective rshares change over time on a single piece of content. For a small account, it is not that difficult to imagine a scenario where they vote and later want to change their vote (say from 80% to 100%) but because the vote dust threshold was updated they actually allocate less rshares.

Granted, anything close to the dust threshold will be impacted heavily by it but will also have very little impact on the end payout of a post, but it could lead to bad UX. And is significantly more complicated if that value can change.

This is all good theory, but let's figure out what the smallest amount of changes we need to make is.

What is currently implemented is: `effective_rshares = max( 0, rshares - dust_threshold )`.

Personally, I am 100% comfortable releasing this.

We can also release: `effective_rshares = (rshares > dust_threshold) ? rshares - dust_threshold/2 : 0`

From a technical perspective I am comfortable with it, but I think it is less intuitive of a solution to an outside observer.

If we can avoid making this a witness parameter, I think we should. We avoid a lot of complication by doing so.

The next question (if we don't have a witness parameter) is this: is the current dust threshold a good value?

The current value is 50,000,000 rshares. With a 100% vote consuming 2% voting power, the VESTS breakpoint for an account is 2,500,000,000 (or 2500 VESTS). That is equivalent to 1.219 SP. (At the current conversion rate of 487.433 STEEM/MV)

With the current reward pool, that vote would reward around 0.000108 STEEM. My gut was that a tenth of a satoshi was probably around the right value. I would also be comfortable setting the threshold to 500,000,000 (0.001 STEEM) and then subtracting half the threshold.

Even at that value, an account now needs ~ 12.19 STEEM, which is not unreasonable right now. However, if STEEM goes to $10, that could be problematic.

With this in mind, I want to err on the side of caution and not increase the vote dust threshold.

**Comment by iamsmooth** [comment]
I like how this discussion went.

I would point out that I have a similar suggestion (for virtually identical reasons) on the minimum comment payout. That is, not only have a threshold to reach before payout but also a cost to pay (in rshares) per comment. That belongs on another issue though  (don't know if there is one).

**Comment by mvandeberg** [comment]
I am going to submit the PR for this with the same vote dust threshold.

**Comment by holgern** [comment]
In the merged PR #1934, 
 ```
   if( db.has_hardfork( STEEM_HARDFORK_0_20__1764 ) )
   {
      info->abs_rshares -= STEEM_VOTE_DUST_THRESHOLD;
      info->abs_rshares = std::max( int64_t(0), info->abs_rshares );
   }
   if( db.has_hardfork( STEEM_HARDFORK_0_14__259 ) )
   {
      FC_ASSERT( info->abs_rshares > STEEM_VOTE_DUST_THRESHOLD || vote_weight == 0, "Voting weight is too small, please accumulate more voting power or steem power." );
   } 
 ```
info->abs_rshares must now be 2*STEEM_VOTE_DUST_THRESHOLD to sucessfully cast a vote. Should the second if not a else if (libraries/chain/steem_evaluator.cpp, line 1298)?
 ```
   if( db.has_hardfork( STEEM_HARDFORK_0_20__1764 ) )
   {
      info->abs_rshares -= STEEM_VOTE_DUST_THRESHOLD;
      info->abs_rshares = std::max( int64_t(0), info->abs_rshares );
   }
  else if( db.has_hardfork( STEEM_HARDFORK_0_14__259 ) )
   {
      FC_ASSERT( info->abs_rshares > STEEM_VOTE_DUST_THRESHOLD || vote_weight == 0, "Voting weight is too small, please accumulate more voting power or steem power." );
   } 
 ```

**Comment by mvandeberg** [comment]
While working on #2539 I saw that there still exists an assertion that `abs_rshares > 0`. Part of the purpose of this change is to not reject votes with low rshares as it results in a poor user experience. I am fixing the problem in my work for #2539
